spec/controllers/courses_controller_spec.rb

```diff
@@ -1,6 +1,85 @@
 describe CoursesController do
   let(:instructor) { create :instructor }
   let(:course) { instructor.course }
+
+  describe 'role switching methods' do
+    let(:end_user) { create :end_user }
+    let(:subject) do
+      post_as instructor, :switch_role, params: { 'id' => course.id, 'effective_user_login' => end_user.user_name }
+    end
+    describe '#switch_role' do
+      let(:second_instructor) { create :instructor }
+      let(:temp_user) { create :end_user }
+
+      it 'fails when no username is provided ' do
+        post_as instructor, :switch_role, params: { 'id' => course.id }
+        expect(response).to have_http_status(:not_found)
+      end
+      it 'fails when an invalid username is provided' do
+        User.delete(temp_user)
+        post_as instructor, :switch_role, params: { 'id' => course.id, 'effective_user_login' => temp_user.user_name }
+        expect(response).to have_http_status(:not_found)
+      end
+      it 'fails when switching role as the current instructor' do
+        end_user.roles << instructor
+        expect(subject).to have_http_status(:unprocessable_entity)
+      end
+      it 'fails when switching role as another instructor in the course' do
+        end_user.roles << second_instructor
+        expect(subject).to have_http_status(:unprocessable_entity)
+      end
+      context 'when switching to a student' do
```

Comment:
Also check for the response status and `session[:user_name]` here (and for the TA tests below)